### ISSUE TITLE
NullPointerException in DB2Database when constraint schema is unavailable

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#14530]
A customer reported an exception in `DB2Database`, which might be related to specific grants being given to a `UNIQUE INDEX` backing a `PRIMARY KEY CONSTRAINT` in a table:

```
java.lang.RuntimeException: java.lang.RuntimeException: java.lang.NullPointerException: Cannot invoke "org.jooq.meta.SchemaDefinition.getDatabase()" because "schema" is null
    at org.jooq.meta.AbstractDatabase.onError (AbstractDatabase.java:3779)
    at org.jooq.meta.AbstractDatabase.getRelations (AbstractDatabase.java:2791)
    at org.jooq.meta.DefaultColumnDefinition.getPrimaryKey (DefaultColumnDefinition.java:154)
    at org.jooq.meta.AbstractTableDefinition.getPrimaryKey (AbstractTableDefinition.java:96)
    at org.jooq.codegen.JavaGenerator.generateTable (JavaGenerator.java:5872)
    at org.jooq.codegen.JavaGenerator.generateTables (JavaGenerator.java:5851)
    at org.jooq.codegen.JavaGenerator.generate (JavaGenerator.java:652)
    at org.jooq.codegen.JavaGenerator.generate (JavaGenerator.java:607)
    at org.jooq.codegen.JavaGenerator.generate0 (JavaGenerator.java:519)
    at org.jooq.codegen.AbstractGenerator.generate (AbstractGenerator.java:187)
    at org.jooq.codegen.JavaGenerator.generate (JavaGenerator.java:227)
    at org.jooq.codegen.GenerationTool.run0 (GenerationTool.java:942)
    at org.jooq.codegen.GenerationTool.run (GenerationTool.java:240)
    at org.jooq.codegen.GenerationTool.generate (GenerationTool.java:235)
    at org.jooq.codegen.maven.Plugin.execute (Plugin.java:216)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    at org.codehaus.classworlds.Launcher.main (Launcher.java:47)
Caused by: java.lang.RuntimeException: java.lang.NullPointerException: Cannot invoke "org.jooq.meta.SchemaDefinition.getDatabase()" because "schema" is null
    at org.jooq.meta.AbstractDatabase.onError (AbstractDatabase.java:3779)
    at org.jooq.meta.AbstractDatabase.getRelations0 (AbstractDatabase.java:3119)
    at org.jooq.meta.AbstractDatabase.lambda$getRelations$18 (AbstractDatabase.java:2791)
    at org.jooq.meta.AbstractDatabase.onError (AbstractDatabase.java:3753)
    at org.jooq.meta.AbstractDatabase.getRelations (AbstractDatabase.java:2791)
    at org.jooq.meta.DefaultColumnDefinition.getPrimaryKey (DefaultColumnDefinition.java:154)
    at org.jooq.meta.AbstractTableDefinition.getPrimaryKey (AbstractTableDefinition.java:96)
    at org.jooq.codegen.JavaGenerator.generateTable (JavaGenerator.java:5872)
    at org.jooq.codegen.JavaGenerator.generateTables (JavaGenerator.java:5851)
    at org.jooq.codegen.JavaGenerator.generate (JavaGenerator.java:652)
    at org.jooq.codegen.JavaGenerator.generate (JavaGenerator.java:607)
    at org.jooq.codegen.JavaGenerator.generate0 (JavaGenerator.java:519)
    at org.jooq.codegen.AbstractGenerator.generate (AbstractGenerator.java:187)
    at org.jooq.codegen.JavaGenerator.generate (JavaGenerator.java:227)
    at org.jooq.codegen.GenerationTool.run0 (GenerationTool.java:942)
    at org.jooq.codegen.GenerationTool.run (GenerationTool.java:240)
    at org.jooq.codegen.GenerationTool.generate (GenerationTool.java:235)
    at org.jooq.codegen.maven.Plugin.execute (Plugin.java:216)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    at org.codehaus.classworlds.Launcher.main (Launcher.java:47)
Caused by: java.lang.NullPointerException: Cannot invoke "org.jooq.meta.SchemaDefinition.getDatabase()" because "schema" is null
    at org.jooq.meta.AbstractConstraintDefinition.<init> (AbstractConstraintDefinition.java:52)
    at org.jooq.meta.DefaultCheckConstraintDefinition.<init> (DefaultCheckConstraintDefinition.java:49)
    at org.jooq.meta.DefaultCheckConstraintDefinition.<init> (DefaultCheckConstraintDefinition.java:45)
    at org.jooq.meta.db2.DB2Database.loadCheckConstraints (DB2Database.java:334)
    at org.jooq.meta.AbstractDatabase.lambda$getRelations0$27 (AbstractDatabase.java:3119)
    at org.jooq.meta.AbstractDatabase.onError (AbstractDatabase.java:3753)
    at org.jooq.meta.AbstractDatabase.getRelations0 (AbstractDatabase.java:3119)
    at org.jooq.meta.AbstractDatabase.lambda$getRelations$18 (AbstractDatabase.java:2791)
    at org.jooq.meta.AbstractDatabase.onError (AbstractDatabase.java:3753)
    at org.jooq.meta.AbstractDatabase.getRelations (AbstractDatabase.java:2791)
    at org.jooq.meta.DefaultColumnDefinition.getPrimaryKey (DefaultColumnDefinition.java:154)
    at org.jooq.meta.AbstractTableDefinition.getPrimaryKey (AbstractTableDefinition.java:96)
    at org.jooq.codegen.JavaGenerator.generateTable (JavaGenerator.java:5872)
    at org.jooq.codegen.JavaGenerator.generateTables (JavaGenerator.java:5851)
    at org.jooq.codegen.JavaGenerator.generate (JavaGenerator.java:652)
    at org.jooq.codegen.JavaGenerator.generate (JavaGenerator.java:607)
    at org.jooq.codegen.JavaGenerator.generate0 (JavaGenerator.java:519)
    at org.jooq.codegen.AbstractGenerator.generate (AbstractGenerator.java:187)
    at org.jooq.codegen.JavaGenerator.generate (JavaGenerator.java:227)
    at org.jooq.codegen.GenerationTool.run0 (GenerationTool.java:942)
    at org.jooq.codegen.GenerationTool.run (GenerationTool.java:240)
    at org.jooq.codegen.GenerationTool.generate (GenerationTool.java:235)
    at org.jooq.codegen.maven.Plugin.execute (Plugin.java:216)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    at org.codehaus.classworlds.Launcher.main (Launcher.java:47)
```

The error appears because we don't currently filter on `CHECKS.TABSCHEMA` being a part of the input schemata.

**Comment by lukaseder** [comment]
The query is just:

```java
for (CheckRecord r : create()
    .select(
        CHECKS.TABSCHEMA,
        CHECKS.TABNAME,
        CHECKS.CONSTNAME,
        CHECKS.TEXT
    )
    .from(CHECKS)
    .where(CHECKS.TYPE.eq(inline("C")))
    .orderBy(CHECKS.TABSCHEMA, CHECKS.TABNAME, CHECKS.CONSTNAME)
    .fetch(Records.mapping(CheckRecord::new))
) { ... }
```

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.18.0, 3.17.8 (https://github.com/jOOQ/jOOQ/issues/14531), and 3.16.14 (https://github.com/jOOQ/jOOQ/issues/14532)
